### PR TITLE
5192 include in search empty page results

### LINKED ISSUE #5192: When retrieving lists of resources with an _include filter, return values and page sizes are both incorrect
Reproduction:

Create a number of resources with references to different resources.

Submit a search for the resources with an _include filter for the referenced resources

Eg:

```
Task task = new Task();
task.getOwner().setReference(organization.getId().toUnqualifiedVersionless());
// add other task properties here and save task
```

Then conduct a search based on the URL (make sure that there are multiple Tasks that can be returned).

`Task?_count=10&_include=Task%3Aowner` (and any search filters needed)

Returned values are not equal to the submitted count (could be more or less, depending on when the included Organization resources are included). This is expected behaviour.

However, once the list runs out of resources, it'll continue producing "next" pages until no (or < _count) resources are returned, where the number of returned resources *includeS* the _includes

### PR DESCRIPTION
**Pull request by TipzCM**
Fixes #5192

### CODE REVIEW

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/method/ResponsePage.java [hunk 1]
```diff
@@ -19,50 +19,435 @@
...
+				if (myNumTotalResults == null) {
+					if (myPageSize == myResourceList.size() + myOmittedResourceCount - myIncludedResourceCount) {
+						// if the size of the resource list - included resources + omitted resources == pagesize
+						// we have more pages
+						return true;
+					}
+				} else if (myResponseBundleRequest.offset + myNumToReturn < myNumTotalResults) {
+					return true;
+				}
+				break;
```
Comment (michaelabuckley) [inline hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/method/ResponsePage.java:216]: These two look very similar.  Are you sure you don't need to have the same myIncludedResourceCount adjustment in the NONCACHED_OFFSET case?

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/method/ResponsePage.java [hunk 2]
```diff
@@ -19,50 +19,435 @@
...
+							myBundleProvider.getUuid(),
+							myBundleProvider.getNextPageId(),
+							myResponseBundleRequest.getRequestParameters());
+					break;
+				case SAVED_SEARCH:
+					next = RestfulServerUtils.createPagingLink(
+							theLinks,
+							myResponseBundleRequest.requestDetails,
+							mySearchId,
+							myResponseBundleRequest.offset + myNumToReturn,
```
Comment (michaelabuckley) [inline hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/method/ResponsePage.java:258]: I'm confused here - is myNumToReturn just the match entries, or does it also count the includes here?  Do we need the same include adjustment here, and in the others?

File: hapi-fhir-storage/src/test/java/ca/uhn/fhir/rest/server/method/ResponseBundleBuilderTest.java [hunk 3]
```diff
@@ -150,6 +151,55 @@ void testNoLimit(boolean theCanStoreSearchResults) {
 		assertNextLink(bundle, DEFAULT_PAGE_SIZE);
 	}
 
+	@Test
+	public void buildResponseBundle_withIncludeFilterAndFewerResultsThanPageSize_doesNotReturnNextLink() {
```
Comment (michaelabuckley) [inline hapi-fhir-storage/src/test/java/ca/uhn/fhir/rest/server/method/ResponseBundleBuilderTest.java:155]:
```suggestion
	public void buildResponseBundle_withIncludeParamAndFewerResultsThanPageSize_doesNotReturnNextLink() {
```

**Review by michaelabuckley — APPROVED** [review]

File: hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/7_0_0/5192-include-in-search-paging-fix.yaml [hunk 4]
```diff
@@ -0,0 +1,7 @@
+---
+type: fix
+issue: 5192
+title: "Fixed a bug where searches with an _include filter would
+      break paging resulting in 'next' links to blank pages when
+      no more results are available.
```
[the remark this answers is not in the document]
Comment (TipzCM) [reply hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/7_0_0/5192-include-in-search-paging-fix.yaml:6]: udpated

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/api/server/IBundleProvider.java [hunk 5]
```diff
@@ -34,6 +35,14 @@
 
 public interface IBundleProvider {
 
+	/**
+	 * Returns a ResponsePageBuilder for constructing
+	 * pages that return results.
+	 */
+	default ResponsePage.ResponsePageBuilder getResponsePageBuilder() {
```
[the remark this answers is not in the document]
Comment (TipzCM) [reply hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/api/server/IBundleProvider.java:42]: will remove

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/method/ResponsePage.java [hunk 6]
```diff
@@ -19,50 +19,435 @@
...
+		SAVED_SEARCH,
+		/**
+		 * Our unit tests have a tendency to not include all the mocked out
+		 * values, resulting in invalid search pages being returned.
+		 *
+		 * These tests also do not check links, so correctness was never verified.
+		 * This search status should never appear in production.
+		 */
+		UNKNOWN;
 	}
```
[the remark this answers is not in the document]
Comment (TipzCM) [reply hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/method/ResponsePage.java:452]: updated

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/method/ResponsePage.java (hunk 2 above)
Comment (TipzCM) [reply hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/method/ResponsePage.java:258]: `myNumToReturn` is the maximum number to return (as far as i can tell).

This is typically (but not always) the page size.

If you're at the end of your results, and you have a page size of 10, but only 4 resources are being returned, myNumToReturn is 4.

File: hapi-fhir-storage/src/test/java/ca/uhn/fhir/rest/server/method/ResponseBundleBuilderTest.java (hunk 3 above)
Comment (TipzCM) [reply hapi-fhir-storage/src/test/java/ca/uhn/fhir/rest/server/method/ResponseBundleBuilderTest.java:155]: updated

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/method/ResponsePage.java (hunk 1 above)
Comment (TipzCM) [reply hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/method/ResponsePage.java:215]: As far as i can tell, no?

null myNumTotalResults in non-saved searches is assumed to have next pages (so no additional calculations are needed).

I'll take a second look at this monday though

Comment (TipzCM) [reply hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/method/ResponsePage.java]: changed to handle this in both cases

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/PersistedJpaBundleProvider.java [hunk 7]
```diff
@@ -521,7 +535,13 @@ protected List<IBaseResource> toResourceList(ISearchBuilder theSearchBuilder, Li
 		List<IBaseResource> resources = new ArrayList<>();
 		theSearchBuilder.loadResourcesByPid(thePids, includedPidList, resources, false, myRequest);
 
+		// we will send the resource list to our interceptors
+		// this can (potentially) change the results being returned.
+		int precount = resources.size();
 		resources = ServerInterceptorUtil.fireStoragePreshowResource(resources, myRequest, myInterceptorBroadcaster);
+		theResponsePageBuilder.setOmittedResourceCount(precount - resources.size());
```
Comment (michaelabuckley) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/PersistedJpaBundleProvider.java:542]: question: can myOmittedResource count be non-zero before this?  If so, then this should be 
```suggestion
		theResponsePageBuilder.setOmittedResourceCount(theResponsePageBuilder.getOmittedResourceCount() + precount - resources.size());
```

**Review by michaelabuckley — APPROVED** [review]

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/PersistedJpaBundleProvider.java (hunk 7 above)
Comment (TipzCM) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/PersistedJpaBundleProvider.java:542]: probably not. But i've updated the code to allow for this case anyways.

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/provider/r4/ResourceProviderR4Test.java [hunk 8]
```diff
@@ -2651,9 +2650,72 @@ public void testImagingStudyResources() throws Exception {
...
+				assertFalse(bundle.getEntry().isEmpty());
+				count += received;
+			} else {
+				nextUrl = null;
+			}
+		} while (nextUrl != null);
+
+		// verify
+		// we should receive all resources and linked resources
+		assertEquals(total + orgs, count);
```
[the remark this answers is not in the document]
Comment (TipzCM) [reply hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/provider/r4/ResourceProviderR4Test.java:2715]: added

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/api/server/IBundleProvider.java [hunk 9]
```diff
@@ -118,20 +119,49 @@ default Integer getCurrentPageSize() {
 	 * server's processing rules (e.g. _include'd resources, OperationOutcome, etc.). For example,
 	 * if the method is invoked with index 0,10 the method might return 10 search results, plus an
 	 * additional 20 resources which matched a client's _include specification.
+	 * </p>
```
[the remark this answers is not in the document]
Comment (TipzCM) [reply hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/api/server/IBundleProvider.java:122]: updated

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/method/ResponseBundleBuilder.java [hunk 10]
```diff
@@ -91,31 +90,45 @@ private ResponsePage buildResponsePage(ResponseBundleRequest theResponseBundleRe
...
 			pageSize = pagingCalculatePageSize(requestedPage, server.getPagingProvider());
 
-			if (bundleProvider.size() == null) {
+			Integer size = bundleProvider.size();
+			if (size == null) {
 				numToReturn = pageSize;
 			} else {
-				numToReturn = Math.min(pageSize, bundleProvider.size() - theResponseBundleRequest.offset);
+				numToReturn = Math.min(pageSize, size.intValue() - theResponseBundleRequest.offset);
 			}
```
[the remark this answers is not in the document]
Comment (TipzCM) [reply hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/method/ResponseBundleBuilder.java:112]: used ternary op

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/method/ResponseBundleBuilder.java [hunk 11]
```diff
@@ -235,109 +256,19 @@ private BundleLinks buildLinks(ResponseBundleRequest theResponseBundleRequest, R
 				RestfulServerUtils.prettyPrintResponse(server, theResponseBundleRequest.requestDetails),
 				theResponseBundleRequest.bundleType);
 
+		// set self link
```
[the remark this answers is not in the document]
Comment (TipzCM) [reply hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/method/ResponseBundleBuilder.java:259]: removed

File: hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/7_0_0/5192-include-in-search-paging-fix.yaml [hunk 12]
```diff
@@ -0,0 +1,7 @@
+---
+type: fix
+issue: 5192
+title: "Fixed a bug where search Bundles with `include` entries from an _include query parameter might
+      trigger a 'next' link to blank pages when
+      no more results `match` results are available.
+"
```
[the remark this answers is not in the document]
Comment (TipzCM) [reply hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/7_0_0/5192-include-in-search-paging-fix.yaml:7]: updated

**Pull request merged**